### DISCUSSION TITLE
Compile error with latest MSVC (17.4.1) in IntegerCastLoop

### DISCUSSION THREAD

**Discussion by annnei** [duckdb/duckdb#5461] [Q&A] [answered]
I am using duckdb 0.6.0 with downloaded augmentated source in MSVC (from https://github.com/duckdb/duckdb/releases/download/v0.6.0/libduckdb-src.zip).
With the previous MSVC version 17.3.x it worked well to compile this version of duckdb.
But after upgrading to 17.4.1, it complains in IntegerCastLoop and I fail to figure out why. Here is the complaint from the compiler:
```
C:\PROGRA~1\MIB055~1\2022\PROFES~1\VC\Tools\MSVC\1434~1.319\bin\Hostx64\x64\cl.exe  /nologo /TP -DDUCKDB_BUILD_LIBRARY -IC:\repos\os-src\external\DuckDB\. /DWIN32 /D_WINDOWS /GR /EHsc /Zi /O2 /Ob1 /DNDEBUG -MT /wd4244 /wd4267 /wd4200 /wd26451 /wd26495 /D_CRT_SECURE_NO_WARNINGS /utf-8 /bigobj /Zc:__cplusplus -std:c++20 /showIncludes /Foexternal\DuckDB\CMakeFiles\DuckDB.dir\libduckdb-src\duckdb.cpp.obj /Fdexternal\DuckDB\CMakeFiles\DuckDB.dir\DuckDB.pdb /FS -c C:\repos\os-src\external\DuckDB\libduckdb-src\duckdb.cpp
C:\repos\os-src\external\DuckDB\libduckdb-src\duckdb.cpp(36429): error C2672: 'IntegerCastLoop': no matching overloaded function found
  C:\repos\os-src\external\DuckDB\libduckdb-src\duckdb.cpp(36373): note: could be 'bool duckdb::IntegerCastLoop(const char *,duckdb::idx_t,T &,bool)'
  C:\repos\os-src\external\DuckDB\libduckdb-src\duckdb.cpp(36451): note: 'bool duckdb::IntegerCastLoop(const char *,duckdb::idx_t,T &,bool)': could not deduce template argument for 'OP'
```
Do you have any idea what this means and how I can get further?

**Comment by carlopi** [comment, marked as answer]
Hi @annnei, I think I might have a fix for this, could you try to reproduce the same workflow with the using the libduckdb-src.zip that you find inside this zip file ([duckdb-binaries-linux](https://github.com/carlopi/duckdb/suites/10595911697/artifacts/528210385))?

(this is a temporary artefact I generated, with a minor modification on top of current master, question is whether merging has any use)

Thanks!

**Reply by annnei** [reply]
Thanks -- I reproduced the issue locally now and went ahead and replaced our current version of duckdb (0.6.1 now) with your files and that does compile.

**Reply by carlopi** [reply]
The patch (https://github.com/duckdb/duckdb/commit/b338a5a29ce975981eb318e3347a3be65fd390cc) just landed in master, so once CI builds are done this should be fixed on the dev version + next releases.

Thanks for pointing this out.

(I believe in this MSVC is somehow wrong in not being able to deduct the right template parameter, but still it's on us to be compatible with a wide range of compilers ...)

**Reply by annnei** [reply]
How do releases work at duckdb? When can I expect this in a release?

**Reply by carlopi** [reply]
Official line on this is "somewhere in February"

**Reply by annnei** [reply]
Tried v0.7.0 now, and had to add decimal_separator as a last argument to the template argument list to make it compile.
Apparently there has been another change regarding IntegerCastLoop. I believ it is this one:

commit 93d8f6952ecdfdbd7bc60245e003d4ff22fcfa1f
Author: Eero Lihavainen <eero.lihavainen@nitor.com>
Date:   Sat Jan 21 14:21:26 2023 +0200

    handle decimal separator in casting

**Reply by carlopi** [reply]
This is not great, and you are right on where the problem lies. Feel free to file a PR, or I can do it tomorrow.

We should probably set up a CI run to fail on this, so that we can catch problem like this during development.

**Reply by annnei** [reply]
Another side effect that I noticed just now is that duckdb no longer compiles with the slightly older VS version v17.2.6. Haven't looked at what that is yet.

I can make a PR for the IntegerCastLoop tomorrow but have no more time today.

**Reply by annnei** [reply]
https://github.com/duckdb/duckdb/pull/6265

While doing this I learned that it is much faster to use the full source than to compile from the amalgamation version. And with the full source I don't seem to see the MSVC compiler issue. So I'll see if it is an option to compile from a tagged source version rather than from downloaded amalgamation.
